File: SimLuxJS/Performance Tests/Dishwashing/DishExample.py
# This is expected to do the same as the neighbouring DishExample.js.

from enum import Enum
import simpy
from datetime import datetime

class DishExampleSimulation():
	def __init__(self, n, preRinsers, washers, rinsers, driers, enableLogging):
		self.dishLogging = enableLogging
		self.env = simpy.Environment()
		self.preRinsers = simpy.Resource(self.env, capacity=preRinsers)
		self.washers = simpy.Resource(self.env, capacity=washers)
		self.rinsers = simpy.Resource(self.env, capacity=rinsers)
		self.driers = simpy.Resource(self.env, capacity=driers)
		
		nHalf = int(n / 2);
		for i in range(1, nHalf + 1):
			self.env.process(Dish(self, Dish.Type.Plate, i, i % 7 == 0).run())
			self.env.process(Dish(self, Dish.Type.Glass, i, i % 5 == 2).run())
    
	def run(self):
		self.env.run()
    
class Dish():
	class Type(Enum):
		Plate = "Plate"
		Glass = "Glass"
	
	def __init__(self, dishExampleSimulation, dishType, nr, hiddenDirt, isNew=False):
		self.deSim = dishExampleSimulation;
		self.dishType = dishType;
		self.nr = nr;
		self.hiddenDirt = hiddenDirt;
		self.isNew = isNew
    
	def log(self, message):
		if self.deSim.dishLogging:
			print(str(self.deSim.env.now) + ": " + self.dishType.value + " nr. " + str(self.nr) + ("(new) " if self.isNew else " ") + message);
	
	def run(self):
		deSim = self.deSim
		env = deSim.env
		request = deSim.preRinsers.request()
		yield request
		self.log("is being pre-prinsed")
		yield env.timeout(1)
		deSim.preRinsers.release(request)
		request = deSim.washers.request()
		yield request
		self.log("is being washed.")
		if self.dishType == Dish.Type.Glass and not self.isNew and self.nr % 10 == 1:
			# This branch demonstrates how new SimEntities can be added into the running simulation. 
			yield env.timeout(0.4)
			self.log("while being washed, the dish washer found out, that there was another glass inside.")
			newGlass = Dish(deSim, Dish.Type.Glass, self.nr, False, isNew=True)
			newGlass.log("is added to the beginning of the washing pipeline.")
			env.process(newGlass.run())
		yield env.timeout(5)
		deSim.washers.release(request)
		request = deSim.rinsers.request()
		yield request
		self.log("is being rinsed.")
		yield env.timeout(1)
		deSim.rinsers.release(request)
		request = deSim.driers.request()
		yield request
		self.log("is being dried.")
		if (self.hiddenDirt):
			yield env.timeout(1.2)
			deSim.driers.release(request)
			request = deSim.washers.request()
			yield request
			self.log("is being washed again because the drier found some hidden dirt.")
			yield env.timeout(3)
			deSim.washers.release(request)
			request = deSim.rinsers.request()
			yield request
			self.log("is being rinsed again after the hidden dirt had been washed off.")
			yield env.timeout(1)		
			deSim.rinsers.release(request)
			request = deSim.driers.request()
			yield request
			self.log("is being dried again after the hidden dirt had been rinsed.")
		yield env.timeout(2 if self.dishType == Dish.Type.Plate else 3)	# glasses take longer to dry.
		deSim.driers.release(request)
		self.log("is completely clean.")
	
def testPerformance():
	outputFilePath = "Performance Tests/Dishwashing/pythonOutput.csv"
	outputFile = open(outputFilePath, 'w', 1)
	
	def stopTime(startTime):
		return (datetime.now() - startTime).total_seconds()

	repeatitions = 10;

	preRinsersCount = 2;
	washersCount = 4;
	rinsersCount = 2;
	driersCount = 3;
	
	outputFile.write(datetime.now().isoformat() + ",,,,\n")
	outputFile.write("preRinsers: " + str(preRinsersCount) + ",washers: " + str(washersCount) + ",rinsers: " + str(rinsersCount) + ",driers: " + str(driersCount) + ",repeatitions: " + str(repeatitions) + "\n")
	outputFile.write(",,,,\n")
	outputFile.write("logging,n,preparationDuration,runDuration,totalDuration\n")
        
	for logging in [True, False]:
		for n in [100, 1000, 10000, 25000, 50000, 100000, 250000]:
			outputFile.write(",,,,\n");
			
			for simNr in range(repeatitions):
				outputFile.write(str(logging) + "," + str(n))
				preparationStartTime = datetime.now()
				print("starting (logging: " + str(logging) + ", n: " + str(n) + ") at real time: " + preparationStartTime.isoformat())
				sim = DishExampleSimulation(n, preRinsersCount, washersCount, rinsersCount, driersCount, logging)
				preparationDuration = stopTime(preparationStartTime)
				preparationDurationStr = str(preparationDuration)
				print("preparation duration: " + preparationDurationStr + "s")
				outputFile.write("," + preparationDurationStr)
				
				runStartTime = datetime.now()
				sim.run()
				runDuration = stopTime(runStartTime)
				runDurationStr = str(runDuration)
				print("run duration: " + runDurationStr + "s")
				totalDurationStr = str(preparationDuration + runDuration)
				print("Finished at simulated time: " + str(sim.env.now) + " after " + totalDurationStr + " real seconds.")
				print("Total duration: " + totalDurationStr)
				outputFile.write("," + runDurationStr + "," + totalDurationStr + "\n")
	outputFile.close()

testPerformance()

File: SimLuxJS/Performance Tests/Dishwashing/DishExample.js
// Same idea as in examples/DishExample.js, but changed for performance tests.
const {SimLuxJS, SimEntity} = require('../../SimLuxJS.js');

class DishExampleSimulation {
	constructor(n, preRinsers, washers, rinsers, driers, enableLogging) {
		this.dishLogging = enableLogging;
		this.simLuxJS = new SimLuxJS();
		this.preRinsers = this.simLuxJS.createResource(preRinsers);
		this.washers = this.simLuxJS.createResource(washers);
		this.rinsers = this.simLuxJS.createResource(rinsers);
		this.driers = this.simLuxJS.createResource(driers);
		
		let nHalf = n / 2;
		for (let i = 1; i <= nHalf + 1; i++) {
			this.simLuxJS.addSimEntity(new Dish(this, Dish.Type.Plate, i, i % 7 == 0));
			this.simLuxJS.addSimEntity(new Dish(this, Dish.Type.Glass, i, i % 5 == 2));
		}
    }

	async run() {
		await this.simLuxJS.run();
	}
}

class Dish extends SimEntity {
	static Type = {
		Plate: "Plate",
		Glass: "Glass",
	}

	constructor(dishExampleSimulation, dishType, nr, hiddenDirt) {
		super();
		this.deSim = dishExampleSimulation;
		this.dishType = dishType;
		this.nr = nr;
		this.hiddenDirt = hiddenDirt;
	}

	log(message) {
		if (this.deSim.dishLogging) {
			console.log(this.deSim.simLuxJS.getTime() + ": " + this.dishType + " nr. " + this.nr + " " + message);
		}
	}

	async run() {
		let deSim = this.deSim;
		let simLuxJS = deSim.simLuxJS;
		let release = await simLuxJS.waitForResource(deSim.preRinsers);
		this.log("is being pre-prinsed");
		await simLuxJS.advance(1);
		release();
		release = await simLuxJS.waitForResource(deSim.washers);
		this.log("is being washed.");
		if (this.dishType == Dish.Type.Glass && this.nr % 10 == 1) {
			// This branch demonstrates how new SimEntities can be added into the running simulation. 
			await simLuxJS.advance(0.4);
			this.log("while being washed, the dish washer found out, that there was another glass inside.")
			let newGlass = new Dish(deSim, Dish.Type.Glass, this.nr + "(new)", false)
			newGlass.log("is added to the beginning of the washing pipeline.")
			simLuxJS.addSimEntity(newGlass);
		}
		await simLuxJS.advance(5);
		release();
		release = await simLuxJS.waitForResource(deSim.rinsers);
		this.log("is being rinsed.");
		await simLuxJS.advance(1);
		release();
		release = await simLuxJS.waitForResource(deSim.driers);
		this.log("is being dried.");
		if (this.hiddenDirt) {
			await simLuxJS.advance(1.2);
			release();
			release = await simLuxJS.waitForResource(deSim.washers);
			this.log("is being washed again because the drier found some hidden dirt.");
			await simLuxJS.advance(3);
			release();
			release = await simLuxJS.waitForResource(deSim.rinsers);
			this.log("is being rinsed again after the hidden dirt had been washed off.");
			await simLuxJS.advance(1);
			release();
			release = await simLuxJS.waitForResource(deSim.driers);
			this.log("is being dried again after the hidden dirt had been rinsed.");
		}
		await simLuxJS.advance(this.dishType == Dish.Type.Plate ? 2 : 3); // glasses take longer to dry.
		release();
		this.log("is completely clean.");
	}
}

async function testPerformance() {
	const fs = require('node:fs');

	let outputFilePath = "Performance Tests/Dishwashing/nodejsOutput_with_logging_vscode.csv"
	let outputFile = fs.openSync(outputFilePath, "w");
	
	function errorHandler(error) {
		if (error !== null)
			console.error(err);
	}
	
	function appendToOutput(content) {
		fs.writeSync(outputFile, content);
	}
	
	function stopTime(startTime) {
		return (new Date() - startTime) / 1000;
	}

	let repeatitions = 10;

	let preRinsersCount = 2;
	let washersCount = 4;
	let rinsersCount = 2;
	let driersCount = 3;
	
	appendToOutput(new Date().toISOString() + ",,,,\n");
	appendToOutput("preRinsers: " + preRinsersCount + ",washers: " + washersCount + ",rinsers: " + rinsersCount + ",driers: " + driersCount + ",repeatitions: " + repeatitions + "\n");
	appendToOutput(",,,,\n");
	appendToOutput("logging,n,preparationDuration,runDuration,totalDuration\n");

	for (let logging of [true, false]) {
		for (let n of [100, 1000, 10000, 25000, 50000, 100000, 250000]) {
			appendToOutput(",,,,\n");
			
			for (let simNr = 0; simNr < repeatitions; simNr++ ) {
				appendToOutput("" + logging + "," + n);
				let preparationStartTime = new Date();
				console.log("starting (logging: " + logging + ", n: " + n + ") at real time: " + preparationStartTime.toISOString());
				let sim = new DishExampleSimulation(n, preRinsersCount, washersCount, rinsersCount, driersCount, logging)
				preparationDuration = stopTime(preparationStartTime);
				console.log("preparation duration: " + preparationDuration + "s");
				appendToOutput("," + preparationDuration);
				
				let runStartTime = new Date();
				await sim.run();
				let runDuration = stopTime(runStartTime);
				console.log("run duration: " + runDuration + "s");
				let totalDuration = preparationDuration + runDuration
				console.log("Finished at simulated time: " + sim.simLuxJS.getTime() + " after " + totalDuration + " real seconds.");
				console.log("Total duration: " + totalDuration);
				appendToOutput("," + runDuration + "," + totalDuration + "\n");
			}
		}
	}

	fs.closeSync(outputFile);
}

testPerformance();